### PR TITLE
Linter: Warn the user to not to use capture groups

### PR DESCRIPTION
**Pull request by meetmangukiya**
Using capture groups affect the performance of coala, and it should be
discouraged.

Fixes https://github.com/coala/coala/issues/2070

---
This change is [Reviewable](https://reviewable.io/reviews/coala/coala/2895)

### CODE REVIEW

File: coalib/bearlib/abstractions/Linter.py [hunk 1]
```diff
@@ -122,6 +122,15 @@ def __repr__(cls):
...
     class LinterBase(LocalBear, metaclass=LinterMeta):
+        def __init__(self):
+            if options["output_format"] == "regex":
+                options["output_regex"] = re.compile(options["output_regex"])
+                if options["output_regex"].groups > 8:
+                    self.warn("Performance: Using unnecessary capture groups"
+                              " affects the performance of coala. "
+                              "You should use ``(?:<pattern>)`` instead of"
+                              " ``(<pattern>)`` for your regex")
+
```
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: @Makman2 Is this even close to what we discussed a few days ago? I wrote this today, thats why difference may occur. :worried:

Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: OMG... so many test fails already :(

Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: \+ I like the message, just remove the leading `Performance:`-"tag" :)

Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: 1. yes :)
2. python is not doing what you think it does
```
>>> x = re.compile(".(.).")
>>> x.groups
1
```

Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: > >>> x = re.compile(".(.).")
>>> x.groups
1

I think it is doing exactly what I wanted it to do.. One capture group it has, and so it returns `1` :confused: 

From the docs :

> regex.groups
The number of capturing groups in the pattern.

**Comment by Makman2** [comment]
ping @meetmangukiya :)
let's get this done :)

**Comment by meetmangukiya** [comment]
Pong @makman2

**Comment by sils** [comment]
@meetmangukiya can you rebase? Also use closes, not fixes

**Comment by sils** [comment]
@meetmangukiya if you want a review remove the WIP prefix from PR name *before* doing a push, otherwise nobody will look at it

**Comment by meetmangukiya** [comment]
@Makman2 
1. removed the performance tag 
2. added the period.
3. + [ ] What do we have to do about 
> why > 8? :O

**Comment by Makman2** [comment]
because `> 8` is totally wrong :D when the user specifies only 3 valid groups, this condition fails, though it's valid. You need to check `groupindex` and `groups` together:
```
capturing_groups = rgx.groups - len(rgx.groupindex)
if capturing_groups:
    # warn the user
```

**Comment by meetmangukiya** [comment]
ohh.. you're right, I completely forgot that few of them are completely optional if not all.

File: coalib/bearlib/abstractions/Linter.py [hunk 2]
```diff
@@ -57,6 +57,33 @@ def _prepare_options(options):
...
+        }
+
+        no_of_non_named_groups = options["output_regex"].groups
+            - len(options["output_regex"].groupindex)
+        if no_of_non_named_groups:
+            logging.warn("Using unnecessary capture groups"
+                         " affects the performance of coala. "
+                         "You should use ``(?:<pattern>)`` instead of"
+                         " ``(<pattern>)`` for your regex.")
+        else:
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: no else :)

File: coalib/bearlib/abstractions/Linter.py [hunk 3]
```diff
@@ -57,6 +57,33 @@ def _prepare_options(options):
...
+            'additional_info'
+        }
+
+        no_of_non_named_groups = options["output_regex"].groups
+            - len(options["output_regex"].groupindex)
+        if no_of_non_named_groups:
+            logging.warn("Using unnecessary capture groups"
+                         " affects the performance of coala. "
+                         "You should use ``(?:<pattern>)`` instead of"
+                         " ``(<pattern>)`` for your regex.")
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: same here

File: coalib/bearlib/abstractions/Linter.py [hunk 4]
```diff
@@ -57,6 +57,33 @@ def _prepare_options(options):
...
+        else:
+            for capture_group_name in options["output_regex"].groupindex:
+                if capture_group_name not in supported_names:
+                    logging.warn("Using unnecessary capture groups"
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: please rephrase:
```
Unknown named capturing group 'xxx' used. 'xxx-bear' will ignore it.
```
or similar :)

Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: and maybe mention it should be removed if it's not a typo^^

File: coalib/bearlib/abstractions/Linter.py [hunk 5]
```diff
@@ -57,6 +57,33 @@ def _prepare_options(options):
...
+                                 " affects the performance of coala. "
+                                 "You should use ``(?:<pattern>)`` instead of"
+                                 " ``(<pattern>)`` for your regex.")
+                    break
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: no break, warn on every unrecognized group :)

File: coalib/bearlib/abstractions/Linter.py (hunk 3 above)
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: I'm also not sure whether we should print RST to the terminal (I'm really not sure, I know we use it in docs, but double quotes are quite verbose and could irritate). Currently I slightly prefer markdown syntax, but it's up to you :)

Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: Ironically I don't know how to write rst

**Comment by meetmangukiya** [comment]
@makman2 I thought about this and I think we should close this PR. We should not bother the user for a mistake        that the developer of bear made. 

We can have this mmerged after we have some option like `--dev-warnings` .

What are your thoughts?

**Comment by Makman2** [comment]
honestly, coala is a tool for developers, and when putting this message into the debug channel, nobody would read it. I'm more in favor of a warning^^

File: coalib/bearlib/abstractions/Linter.py (hunk 4 above)
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]:
```
Unknown named capturing group 'xxx' used. 'xxx-bear' will ignore it.
```
You assume that the name of capture group and the bear name is same :thinking: ?

File: coalib/bearlib/abstractions/Linter.py (hunk 2 above)
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: I think we remove the whole if else branch.. I'll update the PR, have a look then ;)

**Comment by meetmangukiya** [comment]
How does this look, @Makman2 from design perspective? 
I see the tests are failing.
And theres lack of coverage in function metadata file :O

File: coalib/bearlib/abstractions/Linter.py [hunk 6]
```diff
@@ -68,21 +68,28 @@ def _prepare_options(options):
...
+                known_named_capture_groups += 1
+
+        if unknown_named_capture_groups or no_of_non_named_groups:
+            logging.warn("{} unknown named capture groups are used.\
+                          {} unnamed capture groups are used.\
+                          They are not performance efficient.\
+                          ".format(unknown_named_capture_groups,
+                                   no_of_non_named_groups)
+                         "You should use `(?:<pattern>)` instead of "
+                         "`(<pattern>)` for your regex.")
```
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: ohh... this makes sense

Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: So, I guess in the end I am just copying the code :sweat:

**Comment by meetmangukiya** [comment]
How do I get the lines covered :thinking: ?

**Comment by Makman2** [comment]
you have to write tests

File: tests/bearlib/abstractions/LinterTest.py [hunk 7]
```diff
@@ -6,6 +7,7 @@
 from unittest.mock import ANY, Mock
 from unittest.case import skipIf
 
+from coalib.bearlib.abstractions.Linter import _prepare_options
```
Comment (meetmangukiya) [inline tests/bearlib/abstractions/LinterTest.py]: Remove this import

File: coalib/bearlib/abstractions/Linter.py [hunk 8]
```diff
@@ -57,6 +58,32 @@ def _prepare_options(options):
...
+            'column',
+            'end_line',
+            'end_column',
+            'additional_info'
+        }
+        no_of_non_named_groups = options['output_regex'].groups \
+            - len(options['output_regex'].groupindex)
+
+        if no_of_non_named_groups:
+            logging.warning('Using unnecessary capturing groups '
```
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: @Makman2 I don't think we can get the bearname in the log sadly :disappointed: 

If you can help me out, then it'd great :D

File: coalib/bearlib/abstractions/Linter.py [hunk 9]
```diff
@@ -57,6 +58,32 @@ def _prepare_options(options):
...
+                            'affects the performance of coala. '
+                            "You should use '(?:<pattern>)' instead of "
+                            "'(<pattern>)' for your regex.")
+
+        for capture_group_name in options['output_regex'].groupindex:
+            if capture_group_name not in supported_names:
+                logging.warning("Superfluous capturing group '{}' used. "
```
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: Same here

**Comment by meetmangukiya** [comment]
Finally this PR is ready, after two months :sweat_smile:

File: coalib/bearlib/abstractions/Linter.py (hunk 8 above)
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: create a new parameter `bears` for `_prepare_options` :+1:

File: coalib/bearlib/abstractions/Linter.py [hunk 10]
```diff
@@ -57,6 +58,32 @@ def _prepare_options(options):
...
+            'message',
+            'severity',
+            'line',
+            'column',
+            'end_line',
+            'end_column',
+            'additional_info'
+        }
+        no_of_non_named_groups = options['output_regex'].groups \
+            - len(options['output_regex'].groupindex)
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: would you mind using braces instead of `\` ? :)

Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: Sure :)

File: coalib/bearlib/abstractions/Linter.py [hunk 11]
```diff
@@ -57,6 +58,33 @@ def _prepare_options(options):
...
+            'column',
+            'end_line',
+            'end_column',
+            'additional_info'
+        }
+        no_of_non_named_groups = (options['output_regex'].groups
+                                  - len(options['output_regex'].groupindex))
+
+        if no_of_non_named_groups:
+            logging.warning('{}: Using unnecessary capturing groups '
```
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: @makman2 I'm not quite sure of this.. Don't coala has a centralised `logger` object already?

**Comment by meetmangukiya** [comment]
Out of WIP state

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/coala/coala/2895#-:-KZ5FpHvr9mz44T3sDEL:bmwy7a5)*

File: coalib/bearlib/abstractions/Linter.py (hunk 11 above)
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: the `logging` module is our centralized way of logging^^

File: coalib/bearlib/abstractions/Linter.py [hunk 12]
```diff
@@ -115,6 +142,9 @@ def _prepare_options(options):
 
 
 def _create_linter(klass, options):
+
+    _prepare_options(options, klass.__name__)
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: please provide the class type itself instead of the name, then we have complete access to the bear in `_prepare_options`, which can be potentially used later.

File: coalib/bearlib/abstractions/Linter.py [hunk 13]
```diff
@@ -68,7 +96,6 @@ def _prepare_options(options):
 
             for key, value in options['severity_map'].items():
                 assert_right_type(key, str, 'severity_map key')
-
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: unrelated change

File: tests/bearlib/abstractions/LinterTest.py [hunk 14]
```diff
@@ -55,56 +56,62 @@ def setUp(self):
...
+            linter('some-executable', invalid_arg=88,
+                   ABC=2000)(object)
         self.assertEqual(
             str(cm.exception),
             "Invalid keyword arguments provided: 'ABC', 'invalid_arg'")
 
         with self.assertRaises(ValueError) as cm:
-            linter('some-executable', diff_severity=RESULT_SEVERITY.MAJOR)
+            linter('some-executable', diff_severity=RESULT_SEVERITY.MAJOR)(
+                                                                object)
```
Comment (Makman2) [inline tests/bearlib/abstractions/LinterTest.py]: indentation looks off, what about
```
linter('some-executable',
       diff_severity=...)(object)
```
?

File: tests/bearlib/abstractions/LinterTest.py [hunk 15]
```diff
@@ -177,24 +184,24 @@ def test_decorator_invalid_parameter_types(self):
...
             linter('some-executable',
                    output_format='corrected',
-                   result_message=list())
+                   result_message=list())(object)
 
         with self.assertRaises(TypeError) as cm:
             linter('some-executable',
                    output_format='corrected',
-                   diff_severity=999888777)
+                   diff_severity=999888777)(object)
```
Comment (Makman2) [inline tests/bearlib/abstractions/LinterTest.py]: ah but you should maybe use the test base classes defined in the test-suite, `object` has problems. If you need an empty class, you can use `EmptyTestLinter`. There are a few others around if needed.

File: tests/bearlib/abstractions/LinterTest.py [hunk 16]
```diff
@@ -856,3 +863,24 @@ def create_arguments(filename, file, config_file, some_value_B):
...
+
+    def test_capture_groups_warnings(self):
+        logger = logging.getLogger()
+        with self.assertLogs(logger, 'WARNING') as cm:
+            @linter('some-executable',
+                    use_stdout=True,
+                    output_format='regex',
+                    output_regex=r'(?P<not_supported_name>)\d+(\w+)')
+            class SomeBear:
+                pass
```
Comment (Makman2) [inline tests/bearlib/abstractions/LinterTest.py]: you should be consistent and use functional style like the other tests^^

File: tests/bearlib/abstractions/LinterTest.py [hunk 17]
```diff
@@ -856,3 +863,24 @@ def create_arguments(filename, file, config_file, some_value_B):
...
+        self.assertEqual(cm.output, [
+            'WARNING:root:SomeBear: Using unnecessary capturing groups '
+            'affects the performance of coala. '
+            "You should use '(?:<pattern>)' instead of "
+            "'(<pattern>)' for your regex.",
+
+            'WARNING:root:SomeBear: Superfluous capturing group'
+            " 'not_supported_name' used. Is this a typo? If not, consider"
+            " removing the capturing group to improve coala's "
+            'performance.'])
```
Comment (Makman2) [inline tests/bearlib/abstractions/LinterTest.py]: please be consistent with strings, either end always with a space (which I prefer) or start everywhere with a space, don't mix both styles.

Comment (Makman2) [inline tests/bearlib/abstractions/LinterTest.py]: Also you could actually try to split up the commit to address both issues separately, this should be fairly easy with `git reset HEAD~1; git add -p`

**Comment by sils** [comment]
marking WIP as per review from @Makman2

File: coalib/bearlib/abstractions/Linter.py (hunk 13 above)
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: Done

File: coalib/bearlib/abstractions/Linter.py (hunk 11 above)
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: Ok :+1:

File: coalib/bearlib/abstractions/Linter.py [hunk 18]
```diff
@@ -17,7 +18,7 @@
 from coalib.settings.FunctionMetadata import FunctionMetadata
 
 
-def _prepare_options(options):
+def _prepare_options(options, bear_name):
```
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: Added

File: tests/bearlib/abstractions/LinterTest.py (hunk 16 above)
Comment (meetmangukiya) [inline tests/bearlib/abstractions/LinterTest.py]: Can't use `EmptyLinterBear` in here, it needs to be defined right below the decorator it seems

File: tests/bearlib/abstractions/LinterTest.py (hunk 17 above)
Comment (meetmangukiya) [inline tests/bearlib/abstractions/LinterTest.py]: Splitted in current push :) That was a new thing to learn, I always thought a _wrong_ squash means a lot of work, if wanted to split the commits, though never faced such a situation.

File: coalib/bearlib/abstractions/Linter.py [hunk 19]
```diff
@@ -1,5 +1,6 @@
 from contextlib import contextmanager
 from functools import partial, partialmethod
```
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: commit message: You should add `anonymous capturing groups` ->
```
Linter: Warn on using anonymous capture groups
```

Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: If you want you can also copy the commit body from the other commit to explain why this change is made :+1:

File: tests/bearlib/abstractions/LinterTest.py [hunk 20]
```diff
@@ -872,12 +872,17 @@ def test_capture_groups_warnings(self):
             @linter('some-executable',
                     use_stdout=True,
                     output_format='regex',
-                    output_regex=r'(\w+)')
+                    output_regex=r'(?P<not_supported_name>)\d+(\w+)')
```
Comment (Makman2) [inline tests/bearlib/abstractions/LinterTest.py]: I think you've meant `(?P<not_supported_name>\d+), but that doesn't matter for the test :)

**Comment by Makman2** [comment]
ack 1844193

**Comment by Makman2** [comment]
But looks very good now, we should merge this on the next iteration :+1: :D

File: coalib/bearlib/abstractions/Linter.py [hunk 21]
```diff
@@ -17,12 +18,14 @@
...
 
-def _prepare_options(options):
+def _prepare_options(options, bear_class):
     """
     Prepares options for ``linter`` for a given options dict in-place.
 
     :param options:
         The options dict that contains user/developer inputs.
+    :param bear_class:
+        The Bear ``class`` which is being decorated by ``linter``
```
Comment (meetmangukiya) [inline coalib/bearlib/abstractions/Linter.py]: ugh... I'll add it now.

**Comment by meetmangukiya** [comment]
*[tests/bearlib/abstractions/LinterTest.py, line 875 at r2](https://reviewable.io:443/reviews/coala/coala/2895#-KZ5FQqXQjf_bXL1BDAX-r2-875:-KZBW65hzsgiLR4MLosH:b-pbe08z) ([raw file](https://github.com/coala/coala/blob/1844193c90af44fe9c3507c2f3f95f8f4540ebbe/tests/bearlib/abstractions/LinterTest.py#L875)):*
<details><summary><i>Previously, Makman2 (Mischa Krüger) wrote…</i></summary><blockquote>

I think you've meant `(?P<not_supported_name>\d+)`, but that doesn't matter for the test :)
</blockquote></details>

No I meant this I guess, but nm.. :/ 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/coala/coala/2895)*

**Comment by sims1253** [comment]
1. The discussion is getting huge 
2. You have 2 commits. squash them please.

**Comment by meetmangukiya** [comment]
@sims1253, @Makman2 asked me to create two commits

**Comment by sims1253** [comment]
yup saw that. thats why I remove my comment :)

**Comment by Makman2** [comment]
ack b97dfba, ecea2d1

**Comment by Makman2** [comment]
@meetmangukiya rebase, then we gonna merge 🎉

**Comment by meetmangukiya** [comment]
On it!

**Comment by meetmangukiya** [comment]
I hate last moment rebase, I'll have to wait for tests to pass again :(

**Comment by Makman2** [comment]
hehe :)
you can reack your commits if you haven't changed anything :+1:

**Comment by sils** [comment]
reack https://github.com/coala/coala/pull/2895/commits/957b5e61bd128d8a4b18639ce5fbed7060ff971d  https://github.com/coala/coala/pull/2895/commits/3ef1fb774ccb338590ca72bf90ac06f9d61391b1

**Comment by sils** [comment]
@rultor merge
